### PR TITLE
Add "Hibernate" and "Hybrid Sleep" action buttons

### PR DESCRIPTION
**Pull request by hotsezus**
Was trying to add hibernation buttons to easily hibernate my laptop and came to such a solution.
Code automatically checks if user is able to send machine to hibernation (or hybrid sleep) based on hardware ability, settings and permissions too.

Qh-dffLM0KQ

There is just one requirement - edit your polkit permission rules so standart user can send machine to hibernation via DBus.
You can do this by creating:
> /etc/polkit-1/localauthority/50-local.d/com.ubuntu.enable-hibernate.pkla

And adding this:
```
[Re-enable hibernate by default]
Identity=unix-user:*
Action=org.freedesktop.login1.*
ResultActive=yes
```

You can manually invoke CanHibernate, CanHybridSleep, Hibernate and HybridSleep DBus methods using these bash commands:
```
dbus-send --system --print-reply --dest="org.freedesktop.login1" /org/freedesktop/login1 org.freedesktop.login1.Manager.CanHibernate
```
```
dbus-send --system --print-reply --dest="org.freedesktop.login1" /org/freedesktop/login1 org.freedesktop.login1.Manager.CanHybridSleep
```
```
dbus-send --system --print-reply --dest="org.freedesktop.login1" /org/freedesktop/login1 org.freedesktop.login1.Manager.Hibernate boolean:true
```
```
dbus-send --system --print-reply --dest="org.freedesktop.login1" /org/freedesktop/login1 org.freedesktop.login1.Manager.HybridSleep boolean:true
```

Feature is coming with:
1. https://github.com/elementary/switchboard-plug-power/pull/171
2. https://github.com/elementary/default-settings/pull/162

### CODE REVIEW

File: src/Indicator.vala [hunk 1]
```diff
@@ -129,6 +137,26 @@ public class Session.Indicator : Wingpanel.Indicator {
             }
 
             main_grid.add (suspend);
+            
+            try {
+                if(suspend_interface.can_hibernate () == "yes"){
+                    main_grid.add (hibernate);
+                }
+            } catch (GLib.Error e) {
+                stderr.printf ("%s\n", e.message);
```
Comment (danrabbit) [inline src/Indicator.vala]: This seems like it should throw an error dialog and not fail silently

File: src/Indicator.vala [hunk 2]
```diff
@@ -129,6 +137,26 @@ public class Session.Indicator : Wingpanel.Indicator {
             }
 
             main_grid.add (suspend);
+            
+            try {
+                if(suspend_interface.can_hibernate () == "yes"){
```
Comment (danrabbit) [inline src/Indicator.vala]: Please make sure to add spaces around parentheses 
```suggestion
                if (suspend_interface.can_hibernate () == "yes") {
```

**Comment by danrabbit** [comment]
@AlexDerbenew It looks like you'll probably also want to submit a PR to https://github.com/elementary/default-settings for that polkit config

**Comment by AlexDerbenew** [comment]
> Can you please update your branch to pass the linter?
> 
> I think these errors should probably throw a dialog instead of failing silently. I would suggest using Granite.MessageDialog with some user readable text and then including the error message with https://valadoc.org/granite/Granite.MessageDialog.show_error_details.html

Have you though about using something like [uncrustify](https://github.com/uncrustify/uncrustify)  as git precommit hooks in ElemenatryOS projects?

**Comment by danrabbit** [comment]
@AlexDerbenew If you can resolve conflicts with master and the review comments, I'd like to re-review this :)

**Comment by AlexDerbenew** [comment]
> @AlexDerbenew If you can resolve conflicts with master and the review comments, I'd like to re-review this :)

Yes. I'll definetly do the whole things that need to be done with this feature in several days

File: src/Indicator.vala [hunk 3]
```diff
@@ -126,6 +134,25 @@ public class Session.Indicator : Wingpanel.Indicator {
...
 
             main_grid.add (suspend);
+
+            try {
+                if (suspend_interface.can_hibernate () == "yes") {
+                    main_grid.add (hibernate);
+                }
+            } catch (GLib.Error e) {
+                warning ("CanHibernate Error: %s", e.message);
+                hibernate.set_sensitive (false);
```
Comment (danrabbit) [inline src/Indicator.vala]: Not sure it makes sense to set this widget as insensitive if it's never packed in the first place

File: src/Indicator.vala [hunk 4]
```diff
@@ -126,6 +134,25 @@ public class Session.Indicator : Wingpanel.Indicator {
...
+            }
+
+            try {
+                if (suspend_interface.can_hybrid_sleep () == "yes") {
+                    main_grid.add (hybrid_sleep);
+                }
+            } catch (GLib.Error e) {
+                warning ("CanHybridSleep Error: %s", e.message);
+                hybrid_sleep.set_sensitive (false);
```
Comment (danrabbit) [inline src/Indicator.vala]: Same as above. If the widget is never packed into the container, probably doesn't make a difference if it's sensitive

**Comment by AlexDerbenew** [comment]
It seems like we should place new polkit rule in '/etc/polkit-1/localauthority/10-vendor.d/' directory and it should look like this
```
[Re-enable hibernate by default]
Identity=unix-user:*
Action=org.freedesktop.login1.hibernate;org.freedesktop.login1.handle-hibernate-key;org.freedesktop.login1;org.freedesktop.login1.hibernate-multiple-sessions;org.freedesktop.login1.hibernate-ignore-inhibit
ResultActive=yes
```

**Comment by AlexDerbenew** [comment]
Also, i'm thinking of adding some code to [switchboard-plug-power](https://github.com/elementary/switchboard-plug-power), so system can be consistent with all properties matching each other

**Comment by AlexDerbenew** [comment]
@danrabbit how do you look at adding an additional dialog box that opens when you click on the hibernation and hybrid sleep buttons?  Or maybe I should add hibernation and hybrid sleep buttons from power options list and add them to the menu that opens when you click on the shutdown button?

**Comment by AlexDerbenew** [comment]
Going to rework this PR today on latest daily build with redesigned UX.
I think it would be better if i leave "Hibernate" and "Hybrid Sleep" buttons where they are but add suspend dialog for both of them so user will have to confirm action

**Comment by AlexDerbenew** [comment]
Alright. Force pushed an updated version of this feature to get rid of multiple unnecessary commits.

Updated version asks user to confirm action. It lets user to cancel action if it was misclick and helps to understand what will happen with device after entering selected mode.
Hibernate Dialog
Hybrid Sleep Dialog

By the way, I had to force synchronous initialization of login dbus interface so i can use interface in the ui building process. I'm sure it's ok, but just mentioning it

**Comment by cassidyjames** [comment]
So I believe Hibernate was disabled in Ubuntu was because it does not work with Secure Boot and the kernel's Lockdown mode. Since we want to support secure boot and don't necessarily want differing experiences on devices with and without secure boot, I don't know that we want to pursue re-enabling hibernation downstream from Ubuntu, until that can be resolved—which is being worked on, but not yet complete.

See this blog post from Matthew Garrett for more info: https://mjg59.dreamwidth.org/55845.html

**Comment by hotsezus** [comment]
Btw, i need someone who can check what output `CanHibernate` DBus method returns on system with secure boot enabled and hibernation setup

**Comment by jeremypw** [comment]
> Btw, i need someone who can check what output CanHibernate DBus method returns on system with secure boot enabled and hibernation setup

Copied request to Slack for greater visibility.

**Comment by jeremypw** [comment]
Copypasted from elementary/wingpanel-indicator-session#102 (comment) for context:

> So I believe Hibernate was disabled in Ubuntu was because it does not work with Secure Boot and the kernel's Lockdown mode. Since we want to support secure boot and don't necessarily want differing experiences on devices with and without secure boot, I don't know that we want to pursue re-enabling hibernation downstream from Ubuntu, until that can be resolved—which is being worked on, but not yet complete.

See this blog post from Matthew Garrett for more info: https://mjg59.dreamwidth.org/55845.html

**Comment by jeremypw** [comment]
Closing this as there are concerns that restoring from these suspended states is always reliable and one of the related PRs is already closed.  Thanks for your contribution anyway.

**Comment by hotsezus** [comment]
[As I said here](https://github.com/elementary/wingpanel-indicator-session/pull/102#issuecomment-793808104), I wanted to checked what is happening with my code on pc with secure boot enabled.

Found out that both 

> dbus-send --system --print-reply --dest="org.freedesktop.login1" /org/freedesktop/login1 org.freedesktop.login1.Manager.CanHibernate

 and
> dbus-send --system --print-reply --dest="org.freedesktop.login1" /org/freedesktop/login1 org.freedesktop.login1.Manager.CanHybridSleep

return
> string "na"

 which prevents session indicator from showing 'Hibernate' and 'Hybrid Sleep' buttons

**Pull request closed without merging**